core/sail/api/src/main/java/org/eclipse/rdf4j/common/concurrent/locks/LockManager.java

```diff
@@ -110,40 +110,57 @@ public boolean isActiveLock() {
 
 	/**
 	 * Blocks current thread until the number of active locks has reached zero.
-	 * 
+	 *
 	 * @throws InterruptedException if any thread interrupted the current thread before or while the current thread was
 	 *                              waiting for a notification. The interrupted status of the current thread is cleared
 	 *                              when this exception is thrown.
 	 */
 	public void waitForActiveLocks() throws InterruptedException {
 		long now = -1;
+		boolean checkForChange = false;
 		while (true) {
-			boolean nochange;
-			Set<WeakLockReference> before;
+			boolean nochange = false;
+			Set<WeakLockReference> before = null;
+
 			synchronized (activeLocks) {
-				if (activeLocks.isEmpty())
+				if (activeLocks.isEmpty()) {
 					return;
-				before = new HashSet<>(activeLocks);
+				}
+
+				// The call to releaseAbandoned() further down in the code is only called if "System.currentTimeMillis()
+				// - now >= waitToCollect / 2". We optimize the code so we only create a `before` HashSet if we are
+				// actually going to use it.
+				checkForChange = now > -1 && System.currentTimeMillis() - now >= waitToCollect / 2;
```

Comment:
@jeenbroekstra I did some more testing too see what happens when there is an actual abandoned lock to see if it gets picked up and logged. I found with my old solution it would take a very long time before it was logged. So I decided to rewrite it.